File: benchmarks/fixtures/webapp_rb/auth/tokens.rb
# Token validation and generation.

require_relative '../models/user'
require_relative '../models/session'
require_relative '../config'

class TokenError < StandardError
  # Base exception for token errors.
end

class ExpiredTokenError < TokenError
  # Raised when a token has expired.
end

class InvalidScopeError < TokenError
  # Raised when token scope is insufficient.
end

# Generate a new authentication token for a user.
def generate_token(user, expires_in = Config::TOKEN_EXPIRY)
  payload = "#{user.id}:#{Time.now.utc.iso8601}"
  token = Digest::SHA256.hexdigest("#{payload}:#{Config::SECRET_KEY}")
  Session.create(user: user, token: token, expires_in: expires_in)
  token
end

# Validate a token and return the associated user.
def validate_token(token)
  session = lookup_session(token)
  raise TokenError, 'Invalid token' if session.nil?
  raise ExpiredTokenError, 'Token has expired' if session.expired?

  session.user
end

# Look up a session by its token.
def lookup_session(token)
  Session.find_by_token(token)
end

# Refresh an expiring token.
def refresh_token(old_token)
  user = validate_token(old_token)
  revoke_token(old_token)
  generate_token(user)
end

# Revoke a token, invalidating the session.
def revoke_token(token)
  session = lookup_session(token)
  return false unless session

  session.delete
  true
end

# Revoke all tokens for a user.
def revoke_all_tokens(user)
  sessions = Session.find_all_by_user(user)
  count = 0
  sessions.each do |session|
    session.delete
    count += 1
  end
  count
end

File: benchmarks/fixtures/webapp_php/auth/tokens.php
<?php

namespace App\Auth;

use App\AppError;
use App\Config;
use App\Models\Session;
use App\Models\User;

/**
 * Base exception for token errors.
 */
class TokenError extends AppError
{
}

/**
 * Raised when a token has expired.
 */
class ExpiredTokenError extends TokenError
{
}

/**
 * Generate a new authentication token for a user.
 */
function generate_token(User $user, int $expiresIn = Config::TOKEN_EXPIRY): string
{
    $payload = $user->id . ':' . time();
    $token = hash('sha256', $payload . ':' . Config::SECRET_KEY);
    Session::create($user, $token, $expiresIn);

    return $token;
}

/**
 * Look up a session by its token.
 */
function lookup_session(string $token): ?Session
{
    return Session::findByToken($token);
}

/**
 * Validate a token and return the associated user.
 */
function validate_token(string $token): User
{
    $session = lookup_session($token);
    if ($session === null) {
        throw new TokenError('Invalid token');
    }
    if ($session->expired()) {
        throw new ExpiredTokenError('Token has expired');
    }

    return $session->user;
}

/**
 * Revoke a token, invalidating the session.
 */
function revoke_token(string $token): bool
{
    $session = lookup_session($token);
    if ($session === null) {
        return false;
    }
    $session->delete();

    return true;
}

/**
 * Refresh an expiring token.
 */
function refresh_token(string $oldToken): string
{
    $user = validate_token($oldToken);
    revoke_token($oldToken);

    return generate_token($user);
}
